### PR TITLE
add another case to SROA optimization

### PR DESCRIPTION
**Pull request by WalterBright**
Found a case where the SROA should have sliced a struct into two register variables, but didn't. So I fixed it so it did. There were two problems:

1. a < should have been <=
2. A long initializer to a two-int struct should be done as two int initializers

Added some logging instrumentation.

### CODE REVIEW

**Review by ljmf00 — DISMISSED** [review]

**Comment by p0nce** [comment]
I couldn't run the repro test because the digger output doesn't link.

But I believe this test case should reproduce your bug @WalterBright 

```d
version (D_SIMD) { } else { static assert(0); }

mixin template VectorOps(VectorType, ArrayType: BaseType[N], BaseType, size_t N)
{
    enum Count = N;
    alias Base = BaseType;

    BaseType* ptr() return pure nothrow @nogc
    {
        return array.ptr;
    }

    // Unary operators
    VectorType opUnary(string op)() pure nothrow @safe @nogc
    {
        VectorType res = void;
        mixin("res.array[] = " ~ op ~ "array[];");
        return res;
    }

    // Binary operators
    VectorType opBinary(string op)(VectorType other) pure const nothrow @safe @nogc
    {
        VectorType res = void;
        mixin("res.array[] = array[] " ~ op ~ " other.array[];");
        return res;
    }

    // Assigning a BaseType value
    void opAssign(BaseType e) pure nothrow @safe @nogc
    {
        array[] = e;
    }

    // Assigning a static array
    void opAssign(ArrayType v) pure nothrow @safe @nogc
    {
        array[] = v[];
    }

    void opOpAssign(string op)(VectorType other) pure nothrow @safe @nogc
    {
        mixin("array[] "  ~ op ~ "= other.array[];");
    }

    // Assigning a dyn array
    this(ArrayType v) pure nothrow @safe @nogc
    {
        array[] = v[];
    }

    // Broadcast constructor
    this(BaseType x) pure nothrow @safe @nogc
    {
        array[] = x;
    }

    ref inout(BaseType) opIndex(size_t i) inout pure nothrow @safe @nogc
    {
        return array[i];
    }    
}

// Note: can't be @safe with this signature
Vec loadUnaligned(Vec)(const(BaseType!Vec)* pvec) @trusted
{
    // Since this vector is emulated, it doesn't have alignement constraints
    // and as such we can just cast it.
    return *cast(Vec*)(pvec);
}

private template BaseType(V)
{
    alias typeof( ( { V v; return v; }()).array[0]) BaseType;
}

struct float4
{
    float[4] array;
    mixin VectorOps!(float4, float[4]);
}

alias __m128 = float4;

__m128 _mm_setr_ps (float e3, float e2, float e1, float e0) pure @trusted
{
    float[4] result = [e3, e2, e1, e0];
    return loadUnaligned!(float4)(cast(float*)result.ptr);
}

__m128 _mm_movehdup_ps (__m128 a) pure @trusted
{
    // Generates movshdup since LDC 1.3 with -O1 -mattr=+sse3
    // PERF but GDC never generates it
    a.ptr[0] = a.array[1];
    a.ptr[2] = a.array[3];
    return a;
}

void main()
{
    __m128 A = _mm_movehdup_ps(_mm_setr_ps(1, 2, 3, 4));
//    __m128 A = 0.0;
    float[4] correct = [2.0f, 2, 4, 4 ];
    assert(A.array == correct);
} 
```

A different `loadUnaligned` was called.
Note that with DMD, vectors are emulated in intel-intrinsics because using D_SIMD and core.simd can trigger more DMD bugs (I do try sometimes to go towards enabling D_SIMD because it's actually neat when it works).

**Comment by p0nce** [comment]
> which is what is failing in buildkite

Because it acutally stress the DMD optimizer :)

**Comment by p0nce** [comment]
And indeed the GitHub search doesn't work: https://github.com/AuburnSounds/intel-intrinsics/search?q=_mm_setr_ps 
not showing the _mm_setr_ps definition... but it's right there: https://github.com/AuburnSounds/intel-intrinsics/blob/master/source/inteli/xmmintrin.d#L2492

Probably because that particular funcall is used a lot in that file.

**Comment by WalterBright** [comment]
@p0nce Unfortunately, the example compiles and runs successfully with `-O` and `-O -inline`. But since it doesn't use any SIMD instructions, it won't produce an alignment seg fault anyway.

**Comment by p0nce** [comment]
Builkite says the problem is in **s**mmintrin.d not **p**mmintrin.d

Let me build another test case

**Comment by p0nce** [comment]
@WalterBright 
I think that one will reproduce your bug, as it tests the same _mm function that failed on BuildKite.

```d
version (D_SIMD) { } else { static assert(0); }

mixin template VectorOps(VectorType, ArrayType: BaseType[N], BaseType, size_t N)
{
    enum Count = N;
    alias Base = BaseType;

    BaseType* ptr() return pure nothrow @nogc
    {
        return array.ptr;
    }

    // Unary operators
    VectorType opUnary(string op)() pure nothrow @safe @nogc
    {
        VectorType res = void;
        mixin("res.array[] = " ~ op ~ "array[];");
        return res;
    }

    // Binary operators
    VectorType opBinary(string op)(VectorType other) pure const nothrow @safe @nogc
    {
        VectorType res = void;
        mixin("res.array[] = array[] " ~ op ~ " other.array[];");
        return res;
    }

    // Assigning a BaseType value
    void opAssign(BaseType e) pure nothrow @safe @nogc
    {
        array[] = e;
    }

    // Assigning a static array
    void opAssign(ArrayType v) pure nothrow @safe @nogc
    {
        array[] = v[];
    }

    void opOpAssign(string op)(VectorType other) pure nothrow @safe @nogc
    {
        mixin("array[] "  ~ op ~ "= other.array[];");
    }

    // Assigning a dyn array
    this(ArrayType v) pure nothrow @safe @nogc
    {
        array[] = v[];
    }

    // Broadcast constructor
    this(BaseType x) pure nothrow @safe @nogc
    {
        array[] = x;
    }

    ref inout(BaseType) opIndex(size_t i) inout pure nothrow @safe @nogc
    {
        return array[i];
    }    
}

// Note: can't be @safe with this signature
Vec loadUnaligned(Vec)(const(BaseType!Vec)* pvec) @trusted
{
    // Since this vector is emulated, it doesn't have alignement constraints
    // and as such we can just cast it.
    return *cast(Vec*)(pvec);
}

private template BaseType(V)
{
    alias typeof( ( { V v; return v; }()).array[0]) BaseType;
}

struct int4
{
    int[4] array;
    mixin VectorOps!(int4, int[4]);
}

alias __m128i = int4;

__m128i _mm_setr_epi32 (int e3, int e2, int e1, int e0) pure @trusted
{
    pragma(inline, true);
    int[4] result = [e3, e2, e1, e0];
    return cast(__m128i)( loadUnaligned!(int4)(result.ptr) );
}

/// Insert the 32-bit integer `i` into `a` at the location specified by `imm8[1:0]`.
__m128i _mm_insert_epi32 (__m128i a, int i, const int imm8) pure @trusted
{
    // GDC: nothing special to do, pinsrd generated with -O1 -msse4.1
    // LDC x86: psinrd since LDC 1.1 -O2 with -mattr=+sse4.1
    // LDC arm64: ins.s since LDC 1.8 -O2
    int4 ia = cast(int4)a;
    ia.ptr[imm8 & 3] = i;
    return cast(__m128i)ia; 
}

void main()
{
   __m128i A = _mm_setr_epi32(1, 2, 3, 4);
    int4 C = cast(int4) _mm_insert_epi32(A, 5, 2 + 4);
    int[4] result = [1, 2, 5, 4];
    assert(C.array == result);
} 
```

**Comment by p0nce** [comment]
What changed recently in the library is that I added those pragma(inline, true), I can disable them for DMD if needed.

**Comment by WalterBright** [comment]
Thank you, @p0nce, but I want to fix the problem rather than avoid it!

**Comment by WalterBright** [comment]
> I think that one will reproduce your bug

Success! (Or, should I say, failure, as this does reproduce the bug.) Thank you!

**Comment by p0nce** [comment]
My pleasure!

File: src/dmd/backend/gsroa.d [hunk 1]
```diff
@@ -192,6 +205,7 @@ extern (D) private void sliceStructs_Replace(ref symtab_t symtab, const SymInfo[
                     const n = nthSlice(e);
                     if (getSize(e) == 2 * SLICESIZE)
                     {
+			if (log) { printf("slicing "); elem_print(e); }
```
[the remark this answers is not in the document]
Comment (WalterBright) [reply src/dmd/backend/gsroa.d]: I prefer to do it this way for logging code so it doesn't visually interfere too much with the rest of the code.

**Comment by WalterBright** [comment]
@p0nce it's passing now. Yay!

**Comment by WalterBright** [comment]
@RazvanN7 ping!

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**